### PR TITLE
Add CUDA Kernel for TreeSearch Ngram Blocking

### PR DESCRIPTION
**Pull request by pearlli98**
**Major changes**

1. add availability of cuda kernel for running ngram blocking (both self-blocking and context-blocking), can be activated by the flag `--gpu-beam-blocking True`, functionalities are wrapped in the class `NGramRepeatBlockFunction`
3. all subclasses of `TreeSearch` has new boolean attribute `gpu_beam_blocking` that we use in to `_block_ngrams()`
4. use tensors for the attributes `self.partial_hyps` and `self.context` of `TorchGeneratorAgent` instead of lists
5. add unit tests for gpu ngram blocking 
6. add ninja and protobuf to dependency list to enable CUDA extensions

**Testing steps**

gpu tests: with cuda enabled
1. run `pytest tests/test_transformers.py -k test_beamsearch_blocking_gpu`
2. `pytest tests/test_transformers.py -k test_beamsearch_contextblocking_gpu`

cpu tests:
1. run `pytest tests/test_transformers.py -k test_beamsearch_blocking_cpu`
2. `pytest tests/test_transformers.py -k test_beamsearch_contextblocking_cpu`

**Other information**

1. To run an interactive model with gpu beam blocking, do 
```parlai interactive --model-file "zoo:tutorial_transformer_generator/model" --gpu-beam-blocking True```

**Evaluation**
Have done evaluation on the convai2 teacher task on 3 settings: (1) code on main (2) new code with cpu (3) new code with gpu kernel, results shown below
- Correctness: green check, utterances from all 3 settings are identical.
- Runtime: We are seeing ~10% improvement with the new gpu kernel.

  | main | gpu kernel | cpu
-- | -- | -- | --
average of 10 runs | 697.817s | 620.084s| 689.101s
change | \ | -11.14% | -1.25%

### CODE REVIEW

File: parlai/core/torch_generator_agent.py [hunk 1]
```diff
@@ -956,6 +965,7 @@ def _treesearch_factory(self, device, verbose=False):
             )
         elif method == 'beam':
             return BeamSearch(
+                self.opt['gpu_beam_blocking'],
```
Comment (dexterju27) [inline parlai/core/torch_generator_agent.py]: You have set ```self.gpu_beam_blocking```, why not use it?

File: parlai/core/torch_generator_agent.py [hunk 2]
```diff
@@ -1443,15 +1459,36 @@ def _block_ngrams(
             Source text to grab ngrams from. If None, it uses the current
             hypothesis (i.e. self-blocking).
         """
+        context = None
+        if self.gpu_beam_blocking:
+            if if_context_blocking:
```
Comment (dexterju27) [inline parlai/core/torch_generator_agent.py:1464]: Let's make it if self.gpu_beam_blocking and if_context_blocking?

**Review by dexterju27 — COMMENTED** [review]
The test failure seems to be cased by 
```NameError: name 'NGramRepeatBlock' is not defined``` indicates the CUDA cpp binding function is not handled properly by the CI.

**Comment by pearlli98** [comment]
> is this ready for review? or still WIP?

this is ready for review.

File: parlai/ops/ngram_repeat_block.py [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+    def forward(
+        self,
+        hypothesis,
+        context,
+        lprobs,
+        bsz,
+        step,
+        beam_size,
+        no_repeat_ngram_size,
+        if_context_blocking=False,
```
Comment (klshuster) [inline parlai/ops/ngram_repeat_block.py:50]: is this parameter necessary? I.e., can't we assume that if `context` is passed, we block on it? or does this make the downstream logic easier to handle in the kernel

File: setup.py [hunk 4]
```diff
@@ -9,6 +9,8 @@
 
 from setuptools import setup, find_packages
 
+# from torch.utils.cpp_extension import BuildExtension, CUDAExtension
```
Comment (klshuster) [inline setup.py:12]: nit: can you please delete these commented lines

File: parlai/clib/cuda/ngram_repeat_block_cuda.cpp [hunk 5]
```diff
@@ -0,0 +1,50 @@
+/*
```
Comment (stephenroller) [inline parlai/clib/cuda/ngram_repeat_block_cuda.cpp:1]: Is our code freshly rewritten, or did we get it from fastseq. We need to maintain the original copyright headers (in addition to our own) if that's the case. Both are MIT so it's no problem, but we need to include Copyright (c) Microsoft etc

**Review by stephenroller — COMMENTED** [review]
wow great progress!

File: parlai/clib/cuda/ngram_repeat_block_cuda.cpp (hunk 5 above)
Comment (pearlli98) [reply parlai/clib/cuda/ngram_repeat_block_cuda.cpp:1]: Thanks for pointing this out! We definitely referenced their code but i've also made some significant changes. I think we should add the copyright here? can you let me know what's the right move? not super familiar with this.

File: setup.py (hunk 4 above)
Comment (pearlli98) [reply setup.py:12]: deleted

File: parlai/core/torch_generator_agent.py [hunk 6]
```diff
@@ -1725,7 +1792,8 @@ def select_paths(self, logprobs, prior_scores, current_length) -> _PathSelection
         voc_size = logprobs.size(-1)
 
         # get the backtracking hypothesis id as a multiple of full voc_sizes
-        hyp_ids = best_idxs // voc_size
+        # hyp_ids = best_idxs // voc_size
```
[the remark this answers is not in the document]
Comment (pearlli98) [reply parlai/core/torch_generator_agent.py:1795]: deleted

File: parlai/core/torch_generator_agent.py [hunk 7]
```diff
@@ -1367,11 +1391,14 @@ def set_context(self: TSType, context: torch.LongTensor) -> TSType:
             a LongTensor representing the input context; used for context
             ngram blocking, if supplied
         """
-        self.context = context.tolist()
+        self.context = torch.Tensor(context.tolist()).long()
```
[the remark this answers is not in the document]
Comment (pearlli98) [reply parlai/core/torch_generator_agent.py:1394]: outdated change, reverted to main code.

File: parlai/clib/cuda/ngram_repeat_block_cuda.cpp (hunk 5 above)
Comment (dexterju27) [reply parlai/clib/cuda/ngram_repeat_block_cuda.cpp:1]: As we discussed in our mtg, you can cite the original file and say we adapt from it. And maintain the Microsoft copyright header.

File: parlai/clib/cuda/ngram_repeat_block_cuda_kernel.cu [hunk 8]
```diff
@@ -0,0 +1,112 @@
+/*
+Copyright (c) Facebook, Inc. and its affiliates.
+This source code is licensed under the MIT license found in the
```
Comment (dexterju27) [inline parlai/clib/cuda/ngram_repeat_block_cuda_kernel.cu:3]: And what @pearlli98 mentioned about licensing should apply to this file as well, I suppose?

File: parlai/clib/cuda/ngram_repeat_block_cuda_kernel.cu [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+
+// Ban repeated ngrams of length = 'no_repeat_ngram_size'
+__global__ void banRepeatedTokens(long* __restrict__ hypothesis_ptr, 
+                                  long* __restrict__ context_ptr,
+                                  float* __restrict__ lprobs,
+                                  int current_seq_length, 
+                                  int vocab_size,
+                                  int no_repeat_ngram_size,
+                                  bool if_context_blocking) {
+  auto row = blockIdx.x;
```
Comment (dexterju27) [inline parlai/clib/cuda/ngram_repeat_block_cuda_kernel.cu:25]: Could you add some comments here on what row and col means here?

File: parlai/clib/cuda/ngram_repeat_block_cuda_kernel.cu [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+  }
+
+  auto lprob_start = row * vocab_size;
+  extern __shared__ long tokens_shm[];
+  // each thread writes to shared array
+  tokens_shm[col] = *previous_ngram_ptr; 
+
+  // final thread writes the end of previous ngram array to tokens_shm
+  if (col == blockDim.x - 1) {
+    for (int i=1; i<no_repeat_ngram_size; i++){
```
Comment (dexterju27) [inline parlai/clib/cuda/ngram_repeat_block_cuda_kernel.cu:46]: nit: spacing, you could apply a CPP formatter on this file

File: tests/test_transformers.py [hunk 11]
```diff
@@ -422,7 +422,59 @@ def test_beamsearch_blocking(self):
             assert '34 34' not in text
 
     @pytest.mark.nofbcode
-    def test_beamsearch_contextblocking(self):
+    @testing_utils.skipUnlessGPU
```
Comment (dexterju27) [inline tests/test_transformers.py:425]: Thanks for adding all these tests

**Pull request merged**